library/L1_Peripheral/stm32f10x/pwm.hpp

```diff
@@ -0,0 +1,446 @@
+    };
+
+    inline static const PwmPin_t kB0 = {
+      .pin       = pb0,
+      .registers = TIM3,
+      .channel   = 3,
+      .id        = SystemController::Peripherals::kTimer3
+    };
+
+    inline static const PwmPin_t kB1 = {
+      .pin       = pb1,
+      .registers = TIM3,
+      .channel   = 4,
+      .id        = SystemController::Peripherals::kTimer3
+    };
+
+    inline static const PwmPin_t kB6 = {
+      .pin       = pb6,
+      .registers = TIM4,
+      .channel   = 1,
+      .id        = SystemController::Peripherals::kTimer4
+    };
+
+    inline static const PwmPin_t kB7 = {
+      .pin       = pb7,
+      .registers = TIM4,
+      .channel   = 2,
+      .id        = SystemController::Peripherals::kTimer4
+    };
+
+    inline static const PwmPin_t kB8 = {
+      .pin       = pb8,
+      .registers = TIM4,
+      .channel   = 3,
+      .id        = SystemController::Peripherals::kTimer4
+    };
+
+    inline static const PwmPin_t kB9 = {
+      .pin       = pb9,
+      .registers = TIM4,
+      .channel   = 4,
+      .id        = SystemController::Peripherals::kTimer4
+    };
+  };  // Timer
+
+  /// Constructor for a STM32F10x PWM channel.
+  ///
+  /// @param timer - Reference to a const timer description for this
+  ///        instance of the PWM driver.
+  explicit constexpr Pwm(const PwmPin_t & timer) : pwm_pin_(timer) {}
+
+  void ConfigureFrequency(units::frequency::hertz_t frequency) override
+  {
+    static constexpr auto kUpdateGeneration = bit::MaskFromRange(0);
+
+    auto & system                   = SystemController::GetPlatformController();
+    const auto kPeripheralFrequency = system.GetClockRate(pwm_pin_.id);
+
+    auto period      = 1 / frequency;
+    uint32_t product = kPeripheralFrequency * period;
+
+    pwm_pin_.registers->PSC = GetPrescalarValue(product);
+    pwm_pin_.registers->ARR =
+      static_cast<uint16_t>(product / (pwm_pin_.registers->PSC + 1));
+
+    bit::Register(&pwm_pin_.registers->EGR).Set(kUpdateGeneration).Save();
+  }
+
+  void SetDutyCycle(float duty_cycle) override
+  {
+    // Clamp the duty cycle to make sure it's in the right range
+    const float kClampedDutyCycle = std::clamp(duty_cycle, 0.0f, 1.0f);
+
+    switch (pwm_pin_.channel)
+    {
+    case 1:
+      pwm_pin_.registers->CCR1 =
+        static_cast<uint16_t>(kClampedDutyCycle * pwm_pin_.registers->ARR);
+      break;
+
+    case 2:
+      pwm_pin_.registers->CCR2 =
+        static_cast<uint16_t>(kClampedDutyCycle * pwm_pin_.registers->ARR);
+      break;
+
+    case 3:
+      pwm_pin_.registers->CCR3 =
+        static_cast<uint16_t>(kClampedDutyCycle * pwm_pin_.registers->ARR);
+      break;
+
+    case 4:
+      pwm_pin_.registers->CCR4 =
+        static_cast<uint16_t>(kClampedDutyCycle * pwm_pin_.registers->ARR);
+      break;
+    }
+  }
+
+  float GetDutyCycle() override
+  {
+    return
+      static_cast<float>(pwm_pin_.registers->CCR1 / pwm_pin_.registers->ARR);
```

Comment:
```suggestion
    auto ccr1 = static_cast<float>(pwm_pin_.registers->CCR1);
    auto arr = static_cast<float>(pwm_pin_.registers->ARR);
    return crr1 / arr;
```

You are probably wondering why this is necessary. Because if you divide two ints you get a none float number. meaning, if your value is below 1 it will truncate to 0. So we need to cast both of these into float so we can perform floating point division not int division. Also I gave the two variables the name `ccr1` and `arr`, can you give them variable names that represent what they really are.

Reply:
Just noticed this was wrong since we're doing multi channel now. I'll rewrite this a bit